### PR TITLE
TOUR-12: Add a nice UI to easily create Tour and Steps

### PR DESCRIPTION
**Pull request by mouhb**

### CODE REVIEW

File: application-tour-ui/src/main/resources/TourCode/AjaxActions.xml [hunk 1]
```diff
@@ -0,0 +1,536 @@
...
+  <date>1454671493000</date>
+  <contentUpdateDate>1454671489000</contentUpdateDate>
+  <version>1.1</version>
+  <title/>
+  <defaultTemplate/>
+  <validationScript/>
+  <comment/>
+  <minorEdit>false</minorEdit>
+  <syntaxId>xwiki/2.1</syntaxId>
+  <hidden>false</hidden>
```
Comment (gdelhumeau) [inline application-tour-ui/src/main/resources/TourCode/AjaxActions.xml]: Technical docs must be hidden.

File: application-tour-ui/src/main/resources/TourCode/AjaxActions.xml [hunk 2]
```diff
@@ -0,0 +1,536 @@
...
+        #set($element = $request.getParameter("element"))
+        #set($backdrop = $request.getParameter("backdrop"))
+        #set($content = $request.getParameter("content"))
+        #set($placement = $request.getParameter("placement"))
+        #set($targetPage = $request.getParameter("targetPage"))
+        #set($nbSteps = $request.getParameter("nbSteps"))
+        #set($stepOrder = $request.getParameter("stepOrder"))
+        #set($isNewStep = true)
+        #if ("$!stepOrder" != '') ## Edit
+            #set($stepObj = $bootstrapTourDoc.getObject("TourCode.StepClass", "order", "$stepOrder"))
```
Comment (gdelhumeau) [inline application-tour-ui/src/main/resources/TourCode/AjaxActions.xml]: What if I want to change the order of an existing step? Can I?

File: application-tour-ui/src/main/resources/TourCode/StepSheet.xml [hunk 3]
```diff
@@ -0,0 +1,537 @@
...
+  <date>1454671965000</date>
+  <contentUpdateDate>1454454402000</contentUpdateDate>
+  <version>1.1</version>
+  <title>#if($doc.name == 'StepSheet')Step Sheet#{else}$services.display.title($doc, {'displayerHint': 'default'})#end</title>
+  <defaultTemplate/>
+  <validationScript/>
+  <comment/>
+  <minorEdit>false</minorEdit>
+  <syntaxId>xwiki/2.1</syntaxId>
+  <hidden>false</hidden>
```
Comment (gdelhumeau) [inline application-tour-ui/src/main/resources/TourCode/StepSheet.xml]: Same here.

File: application-tour-ui/src/main/resources/TourCode/AjaxActions.xml [hunk 4]
```diff
@@ -0,0 +1,536 @@
...
+   #set($docRef = $request.getParameter('BSTDocRef'))
+   #set($stepOrder =  $mathtool.toInteger($request.getParameter('stepOrder')))
+   #if ($xwiki.exists($docRef))
+      #set($bootstrapTourDoc = $xwiki.getDocument($docRef))
+      #set($stepObj = $bootstrapTourDoc.getObject("TourCode.StepClass", "order", "$stepOrder"))
+      #if ($stepObj)
+         #set($discard = $bootstrapTourDoc.removeObject($stepObj))
+         #set($discard = $bootstrapTourDoc.save())
+         ## Update other steps order
+         #set($hql = "select distinct step.order from Document doc, doc.object(TourCode.StepClass) as step where doc.fullName='$docRef' order by step.order")
```
Comment (gdelhumeau) [inline application-tour-ui/src/main/resources/TourCode/AjaxActions.xml]: All the steps concerning the same tour are now stored in the same document, so this query is not needed anymore.

File: application-tour-ui/src/main/resources/TourCode/AjaxActions.xml (hunk 2 above)
Comment (gdelhumeau) [inline application-tour-ui/src/main/resources/TourCode/AjaxActions.xml]: I think you should not rely on the "order" field to identify an object, but on the real object number instead.

File: application-tour-ui/src/main/resources/TourCode/AjaxActions.xml (hunk 4 above)
Comment (gdelhumeau) [inline application-tour-ui/src/main/resources/TourCode/AjaxActions.xml]: Instead, do $doc.getObjects(), and sort the objects with $sorttool.sort()

File: application-tour-ui/src/main/resources/TourCode/StepSheet.xml [hunk 5]
```diff
@@ -0,0 +1,537 @@
...
+      new XWiki.widgets.StepPopup({pageURL: bstStepURL + '&amp;BSTDocRef=' + encodeURIComponent(BSTDocRef) + '&amp;stepOrder=' + encodeURIComponent(stepOrder)});            
+   });
+   
+   // Initialize step delete btns
+   $( "body" ).on( "click", ".deleteStep", function() {
+      $('.xdialog-modal-container').remove();
+      var BSTDocRef = $("#BSTDocRef").val();
+      var stepOrder = $(this).attr("rel");
+      new XWiki.widgets.ConfirmationBox({onYes: function() {
+         new XWiki.widgets.Notification("Deleting step ...", 'inprogress');
```
Comment (gdelhumeau) [inline application-tour-ui/src/main/resources/TourCode/StepSheet.xml]: Please use translations here.

File: application-tour-ui/src/main/resources/TourCode/StepSheet.xml [hunk 6]
```diff
@@ -0,0 +1,537 @@
...
+         $.ajax({
+	    url: actionsURL + "&amp;action=deleteStep" + "&amp;BSTDocRef=" + encodeURIComponent(BSTDocRef) + "&amp;stepOrder=" + encodeURIComponent(stepOrder),
+            type: 'POST',
+            success: function(rep, textStatus, jqXHR) {
+               $('#stepsContainer').html(rep);
+            },
+	    complete: function(event, XMLHttpRequest, ajaxOptions) {
+               $(".xnotification-inprogress").remove();
+               new XWiki.widgets.Notification("$services.localization.render('Delete step done!')","done");
```
Comment (gdelhumeau) [inline application-tour-ui/src/main/resources/TourCode/StepSheet.xml]: For safety reasons, please use $escapetool.javascript($services.localization.render('...')).

File: application-tour-ui/src/main/resources/TourCode/StepSheet.xml [hunk 7]
```diff
@@ -0,0 +1,537 @@
...
+      <code>.actionBtn {
+   background: none no-repeat scroll 0 center transparent;
+   display: block;
+   font-size: .8em;
+   padding: .3em .3em .3em 20px;
+   text-transform: uppercase;
+   float:right;
+}
+.addItemBtn {
+   background-image: url("$xwiki.getSkinFile('icons/silk/add.png')");
```
Comment (gdelhumeau) [inline application-tour-ui/src/main/resources/TourCode/StepSheet.xml]: or maybe we should use $services.icon.renderHTML() instead (in the sheet, not in the CSS).

File: application-tour-ui/src/main/resources/TourCode/StepTemplate.xml [hunk 8]
```diff
@@ -0,0 +1,184 @@
...
+  <date>1454452648000</date>
+  <contentUpdateDate>1454452648000</contentUpdateDate>
+  <version>1.1</version>
+  <title>Step Template</title>
+  <defaultTemplate/>
+  <validationScript/>
+  <comment/>
+  <minorEdit>false</minorEdit>
+  <syntaxId>xwiki/2.1</syntaxId>
+  <hidden>false</hidden>
```
Comment (gdelhumeau) [inline application-tour-ui/src/main/resources/TourCode/StepTemplate.xml]: same

File: application-tour-ui/src/main/resources/TourCode/StepsListing.xml [hunk 9]
```diff
@@ -0,0 +1,117 @@
...
+  <date>1454671362000</date>
+  <contentUpdateDate>1454671362000</contentUpdateDate>
+  <version>1.1</version>
+  <title/>
+  <defaultTemplate/>
+  <validationScript/>
+  <comment/>
+  <minorEdit>false</minorEdit>
+  <syntaxId>xwiki/2.1</syntaxId>
+  <hidden>false</hidden>
```
Comment (gdelhumeau) [inline application-tour-ui/src/main/resources/TourCode/StepsListing.xml]: Same :)

File: application-tour-ui/src/main/resources/TourCode/TourSheet.xml [hunk 10]
```diff
@@ -57,7 +57,9 @@
 )))
 {{/html}}
 
-Please edit this page with the object editor to add tour steps.
-## TODO: add a step editor (see http://jira.xwiki.org/browse/TOUR-12)
+## Display tour's steps
+{{html wiki="true"}}&lt;label&gt; Steps&lt;/label&gt;{{/html}}
```
Comment (gdelhumeau) [inline application-tour-ui/src/main/resources/TourCode/TourSheet.xml]: The HTML macro seems to be not necessary.

File: application-tour-ui/src/main/resources/TourCode/StepsListing.xml [hunk 11]
```diff
@@ -0,0 +1,117 @@
...
+  <content>{{velocity}}
+{{html}}
+#set($vdoc = $doc)
+#if (!$BSTDocRef)
+   #set($BSTDocRef = $doc.getFullName())
+#end
+#set($hql = "select distinct step.order, step.element, step.title, step.backdrop, step.placement, step.targetPage from Document doc, doc.object(TourCode.StepClass) as step where doc.fullName='$BSTDocRef' order by step.order")
```
Comment (gdelhumeau) [inline application-tour-ui/src/main/resources/TourCode/StepsListing.xml]: Not needed, all the steps are stored in the same doc.

File: application-tour-ui/src/main/resources/TourCode/AjaxActions.xml [hunk 12]
```diff
@@ -0,0 +1,536 @@
...
+    <className>TourCode.StepClass</className>
+    <guid>928603cf-f71d-4cbb-8aa2-dd502666fcff</guid>
+    <property>
+      <backdrop>0</backdrop>
+    </property>
+    <property>
+      <content>kdfjdsjsdf</content>
+    </property>
+    <property>
+      <element>jjsjdsjd</element>
```
Comment (acotiuga) [inline application-tour-ui/src/main/resources/TourCode/AjaxActions.xml]: What should this value represent ?

**Comment by gdelhumeau** [comment]
Merged, thanks.

**Pull request merged**